src/lib/storage/dictionary_segment.hpp

```diff
@@ -25,55 +27,141 @@ class DictionarySegment : public BaseSegment {
   // returns the first value ID that refers to a value >= the search value
   // returns INVALID_VALUE_ID if all values are smaller than the search value
-  ValueID lower_bound(T value) const;
+  ValueID lower_bound(T value) const {
+    auto forward_iterator = std::lower_bound(_dictionary->begin(), _dictionary->end(), value);
+    if (forward_iterator == _dictionary->end()) {
+      return INVALID_VALUE_ID;
+    }
+    return (ValueID)(forward_iterator - _dictionary->begin());
+  }
 
   // same as lower_bound(T), but accepts an AllTypeVariant
-  ValueID lower_bound(const AllTypeVariant& value) const;
+  ValueID lower_bound(const AllTypeVariant& value) const { return lower_bound(value); }
 
   // returns the first value ID that refers to a value > the search value
   // returns INVALID_VALUE_ID if all values are smaller than or equal to the search value
-  ValueID upper_bound(T value) const;
+  ValueID upper_bound(T value) const {
+    auto forward_iterator = std::upper_bound(_dictionary->begin(), _dictionary->end(), value);
+    if (forward_iterator == _dictionary->end()) {
+      return INVALID_VALUE_ID;
+    }
+    return (ValueID)(forward_iterator - _dictionary->begin());
+  }
 
   // same as upper_bound(T), but accepts an AllTypeVariant
-  ValueID upper_bound(const AllTypeVariant& value) const;
+  ValueID upper_bound(const AllTypeVariant& value) const { return upper_bound(value); }
 
   // return the number of unique_values (dictionary entries)
-  size_t unique_values_count() const;
+  size_t unique_values_count() const { return _dictionary->size(); }
 
   // return the number of entries
-  ChunkOffset size() const override;
+  ChunkOffset size() const override { return _attribute_vector->size(); }
 
   // returns the calculated memory usage
-  size_t estimate_memory_usage() const final;
+  size_t estimate_memory_usage() const final {
+    return _dictionary->size() * sizeof(T) + _attribute_vector->size() * _attribute_vector->width();
+  }
 
  protected:
   std::shared_ptr<std::vector<T>> _dictionary;
   std::shared_ptr<BaseAttributeVector> _attribute_vector;
-};
 
+  std::shared_ptr<std::vector<T>> _initialize_dictionary(const std::shared_ptr<ValueSegment<T>> value_segment) {
+    std::set<T> distinct_values;
+    for (auto value : value_segment->values()) {
+      distinct_values.insert(value);
+    }
+    return std::make_shared<std::vector<T>>(distinct_values.begin(), distinct_values.end());
+  }
+
+  int _get_minimal_number_of_bits_for_dictionary_size(size_t size) {
+    if (size <= std::numeric_limits<uint8_t>::max()) {
+      return 8;
+    } else if (size <= std::numeric_limits<uint16_t>::max()) {
+      return 16;
+    } else if (size <= std::numeric_limits<uint32_t>::max()) {
+      return 32;
+    } else {
+      return 64;
+    }
+  }
+
+  std::shared_ptr<BaseAttributeVector> _create_attribute_vector_with_type(
+      int minimal_bytes, std::shared_ptr<ValueSegment<T>> value_segment, std::shared_ptr<std::vector<T>> dictionary) {
```

Comment:
Be explicit with types: Use [u]int(8|16|32|64)_t instead of int, long, uint etc.

Reply:
All parameters can be const, `value_segment` and `dictionary` can be const shared pointer references.